src/compiler/core.ts

```diff
@@ -92,6 +92,41 @@ module ts {
         return result;
     }
 
+    export function reduce<T>(array: T[], f: (a: T, x: T) => T): T;
+    export function reduce<T, U>(array: T[], f: (a: U, x: T) => U, initial: U): U;
+    export function reduce<T, U>(array: T[], f: (a: any, x: T) => any, initial?: any): any {
+        var result = initial;
+        if (array) {
+            for (var i = 0, l = array.length; i < l; i++) {
+                if (i === 0 && arguments.length <= 2) {
+                    result = array[i];
+                }
+                else {
+                    result = f(result, array[i]);
+                }
+            }
+        }
+        return result;
+    }

```

Comment:
Also, it might be worth creating a new `reduce` function like `reduceWithInitial`, or `reduce1`.